cpp/src/parquet/arrow/reader.h

```diff
@@ -191,13 +191,53 @@ class PARQUET_EXPORT FileReader {
   ///
   /// \returns error Status if either row_group_indices or column_indices
   ///     contains an invalid index
+  /// \deprecated Deprecated in future release. Use arrow::Result version instead.
+  ARROW_DEPRECATED("Deprecated in future release. Use arrow::Result version instead.")
   ::arrow::Status GetRecordBatchReader(const std::vector<int>& row_group_indices,
                                        const std::vector<int>& column_indices,
                                        std::shared_ptr<::arrow::RecordBatchReader>* out);
+
+  /// \brief Return a RecordBatchReader of row groups selected from
+  /// row_group_indices, whose columns are selected by column_indices.
+  ///
+  /// Note that the ordering in row_group_indices and column_indices
+  /// matter. FileReaders must outlive their RecordBatchReaders.
+  ///
+  /// \param row_group_indices which row groups to read (order determines read order).
+  /// \param column_indices which columns to read (order determines output schema).
+  ///
+  /// \returns error Result if either row_group_indices or column_indices
+  ///     contains an invalid index
+  virtual ::arrow::Result<std::shared_ptr<::arrow::RecordBatchReader>>
+  GetRecordBatchReaderSharedPtr(const std::vector<int>& row_group_indices,
```

Comment:
```suggestion
  GetRecordBatchReader(const std::vector<int>& row_group_indices,
```

Can we avoid adding `SharedPtr` to be consistent?

Reply:
cc @kou .